### ISSUE TITLE
2019-11-20 tc010 M1SD - test case variant with 3 profiles (SDEIDs) with duplicate records, i.e. exact same TIME and differing CONCENTRATION values - computation engine fails unacceptably

### ISSUE THREAD

**Issue by tensfeldt** [tensfeldt/openNCA#38]
Behavior in this case is that the computation engine should report the issue, report what records are impacted by it and generate missing values for all of the expected parameters for the affected PROFILE but not crash and must continue to process and generate parameters for all subsequent profiles.

The following table shows the duplicated TIME records/values for 3 SDEIDs:

```
     PKDATAROWID     SDEID   SUBJID PKPTMS PKATPD TREATXT PERIOD  PKCNCN
170           27 908214575 10011002    0.0  0.000   10 mg      3  0.0000
171           28 908214575 10011002    0.5  0.600   10 mg      3  6.2900
172           29 908214575 10011002    1.0  1.000   10 mg      3  7.9300
173           30 908214575 10011002    1.5  1.500   10 mg      3  7.6500
174           31 908214575 10011002    2.0  2.000   10 mg      3  8.0100
175           32 908214575 10011002    2.5  2.500   10 mg      3 16.8000
176           33 908214575 10011002    4.0  4.000   10 mg      3 23.1000
177           34 908214575 10011002    8.0  8.000   10 mg      3 16.4000
178           35 908214575 10011002   12.0 12.000   10 mg      3 10.1000
179           36 908214575 10011002   16.0 16.000   10 mg      3  7.5800
1791          37 908214575 10011002   16.0 16.000   10 mg      3 15.1600
180           38 908214575 10011002   24.0 24.000   10 mg      3  6.2800
181           39 908214575 10011002   48.0 48.000   10 mg      3  1.7900
182           40 908214575 10011002   72.0 72.000   10 mg      3  0.5880
118          119 257914734 10011005    0.0  0.000    3 mg      2  0.0000
119          120 257914734 10011005    0.5  0.500    3 mg      2  1.7400
120          121 257914734 10011005    1.0  1.000    3 mg      2  2.9900
121          122 257914734 10011005    1.5  1.500    3 mg      2  2.1500
122          123 257914734 10011005    2.0  2.000    3 mg      2  1.8500
123          124 257914734 10011005    2.5  2.500    3 mg      2  1.5200
124          125 257914734 10011005    4.0  4.017    3 mg      2  2.6700
125          126 257914734 10011005    8.0  8.000    3 mg      2  2.8700
126          127 257914734 10011005   12.0 12.017    3 mg      2  1.7400
127          128 257914734 10011005   16.0 16.000    3 mg      2  1.2300
1271         129 257914734 10011005   16.0 16.000    3 mg      2  2.4600
128          130 257914734 10011005   24.0 24.000    3 mg      2  1.2000
129          131 257914734 10011005   48.0 48.000    3 mg      2  0.6720
130          132 257914734 10011005   72.0 72.000    3 mg      2  0.2240
144          198 949633293 10011008    0.0  0.000    3 mg      2  0.0000
145          199 949633293 10011008    0.5  0.500    3 mg      2  1.3700
146          200 949633293 10011008    1.0  1.000    3 mg      2  1.7500
147          201 949633293 10011008    1.5  1.500    3 mg      2  1.2000
148          202 949633293 10011008    2.0  2.000    3 mg      2  1.2400
149          203 949633293 10011008    2.5  2.533    3 mg      2  1.9400
150          204 949633293 10011008    4.0  4.000    3 mg      2  3.5700
151          205 949633293 10011008    8.0  8.000    3 mg      2  3.6500
152          206 949633293 10011008   12.0 12.000    3 mg      2  2.2300
153          207 949633293 10011008   16.0 16.000    3 mg      2  1.3600
1531         208 949633293 10011008   16.0 16.000    3 mg      2  2.7200
154          209 949633293 10011008   24.0 24.000    3 mg      2  1.1800
155          210 949633293 10011008   48.0 48.000    3 mg      2  0.3400
156          211 949633293 10011008   72.0 72.000    3 mg      2  0.0748
 
```
and the error from the computation engine crash:
```
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
Error in value[[3L]](cond) : 
  Error in FUN(X[[i]], ...): Error in kel: unable to calcualte due to duplicate entires in 'time' vector
For SDEID 257914734
In addition: Warning messages:
1: In run_M1_SD_computation(data = merged_data, map = map_data, method = method,  :
  Flag 'FLGACCEPTTAU' cannot be computed if 'TAUi' is not provided
2: In run_M1_SD_computation(data = merged_data, map = map_data, method = method,  :
 
 Show Traceback
 
 Rerun with Debug
 Error in value[[3L]](cond) : 
  Error in FUN(X[[i]], ...): Error in kel: unable to calcualte due to duplicate entires in 'time' vector
For SDEID 257914734 

```

**Comment by jhhughes256** [comment]
### Update

I can replicate this issue for tc010 by loading the test case unchanged and then running the following code:

```
# Define columns for subset view of data
  sub_cols <- c("PKDATAROWID", "SDEID", "SUBJID", "PKPTMS", "PKATPD", "TREATXT", "PERIOD", "PKCNCN")
# Define IDs to be altered and time to be duplicated
  change_ids <- c(908214575, 257914734, 949633293)
  dupe_time <- 16
# Create new rows for d and change their concentrations by a random amount
  new_d <- d[d$SDEID %in% change_ids & d$PKATPD == dupe_time, ]
  new_d$PKCNCN <- new_d$PKCNCN + runif(3, -1, 1)
# Bind new rows into data and order by PKDATAROWID
  d <- rbind(d, new_d)
  d <- d[order(d$PKDATAROWID), ]
# Determine current max. for row id and the rowids for the IDs to be altered
  max_rowid <- max(d$PKDATAROWID) + 1
  old_rowid <- d[d$SDEID %in% change_ids, "PKDATAROWID"]
# Create a list of unused rowids for the IDs to be altered
  new_rowid <- max_rowid:(max_rowid + nrow(d[d$SDEID %in% change_ids, ]) - 1)
# Apply unused rowids to the altered ID data
  d[d$SDEID %in% change_ids, "PKDATAROWID"] <- new_rowid
# Create new rows for flags, bind and order by PKDATAROWID
  new_flags <- flags[flags$PKDATAROWID %in% new_d$PKDATAROWID, ]
  flags <- rbind(flags, new_flags)
  flags <- flags[order(flags$PKDATAROWID), ]
# Update old flag rowids for the altered ID data
  flags[flags$PKDATAROWID %in% old_rowid, "PKDATAROWID"] <- new_rowid
```

Resulting data below. Error the same as specified above.

```
> d[d$SDEID %in% change_ids, sub_cols]
     PKDATAROWID     SDEID   SUBJID PKPTMS PKATPD TREATXT PERIOD     PKCNCN
118     22291252 257914734 10011005    0.0  0.000    3 mg      2  0.0000000
119     22291253 257914734 10011005    0.5  0.500    3 mg      2  1.7400000
120     22291254 257914734 10011005    1.0  1.000    3 mg      2  2.9900000
121     22291255 257914734 10011005    1.5  1.500    3 mg      2  2.1500000
122     22291256 257914734 10011005    2.0  2.000    3 mg      2  1.8500000
123     22291257 257914734 10011005    2.5  2.500    3 mg      2  1.5200000
124     22291258 257914734 10011005    4.0  4.017    3 mg      2  2.6700000
125     22291259 257914734 10011005    8.0  8.000    3 mg      2  2.8700000
126     22291260 257914734 10011005   12.0 12.017    3 mg      2  1.7400000
127     22291261 257914734 10011005   16.0 16.000    3 mg      2  1.2300000
1271    22291262 257914734 10011005   16.0 16.000    3 mg      2  0.3313965
128     22291263 257914734 10011005   24.0 24.000    3 mg      2  1.2000000
129     22291264 257914734 10011005   48.0 48.000    3 mg      2  0.6720000
130     22291265 257914734 10011005   72.0 72.000    3 mg      2  0.2240000
196     22291266 949633293 10011008    0.0  0.000    3 mg      2  0.0000000
197     22291267 949633293 10011008    0.5  0.500    3 mg      2  1.3700000
198     22291268 949633293 10011008    1.0  1.000    3 mg      2  1.7500000
199     22291269 949633293 10011008    1.5  1.500    3 mg      2  1.2000000
200     22291270 949633293 10011008    2.0  2.000    3 mg      2  1.2400000
201     22291271 949633293 10011008    2.5  2.533    3 mg      2  1.9400000
202     22291272 949633293 10011008    4.0  4.000    3 mg      2  3.5700000
203     22291273 949633293 10011008    8.0  8.000    3 mg      2  3.6500000
204     22291274 949633293 10011008   12.0 12.000    3 mg      2  2.2300000
205     22291275 949633293 10011008   16.0 16.000    3 mg      2  1.3600000
2051    22291276 949633293 10011008   16.0 16.000    3 mg      2  0.8654365
206     22291277 949633293 10011008   24.0 24.000    3 mg      2  1.1800000
207     22291278 949633293 10011008   48.0 48.000    3 mg      2  0.3400000
208     22291279 949633293 10011008   72.0 72.000    3 mg      2  0.0748000
40      22291280 908214575 10011002    0.0  0.000   10 mg      3  0.0000000
41      22291281 908214575 10011002    0.5  0.600   10 mg      3  6.2900000
42      22291282 908214575 10011002    1.0  1.000   10 mg      3  7.9300000
43      22291283 908214575 10011002    1.5  1.500   10 mg      3  7.6500000
44      22291284 908214575 10011002    2.0  2.000   10 mg      3  8.0100000
45      22291285 908214575 10011002    2.5  2.500   10 mg      3 16.8000000
46      22291286 908214575 10011002    4.0  4.000   10 mg      3 23.1000000
47      22291287 908214575 10011002    8.0  8.000   10 mg      3 16.4000000
48      22291288 908214575 10011002   12.0 12.000   10 mg      3 10.1000000
49      22291289 908214575 10011002   16.0 16.000   10 mg      3  7.5800000
4910    22291290 908214575 10011002   16.0 16.000   10 mg      3  7.9304175
50      22291291 908214575 10011002   24.0 24.000   10 mg      3  6.2800000
51      22291292 908214575 10011002   48.0 48.000   10 mg      3  1.7900000
52      22291293 908214575 10011002   72.0 72.000   10 mg      3  0.5880000
```
```

**Comment by rudraya** [comment]
This issue was addressed in 96f6417, please let me know if it still occurs.

**Comment by tensfeldt** [comment]
#38 remains an open issue with tc010 and duplicate time/differing sample concentration tests with commit 96f6417

**Comment by rudraya** [comment]
This issue still occurs because for one of the SDEIDs there happens to be two time points that have different concentration values. As of 96f6417, the code only accounts to remove duplicate time and concentration pair (meaning if both time and concentration are repeated). Please advice on how to deal with the scenario where only time values are repeated but concentration values differ.

**Comment by tensfeldt** [comment]
The circumstance where the duplication occurs with 2 identical time points and 2 concentrations  that differ should be handled by returning missing values, NA, for all pk parameters that are computed, i.e. that profile is invalid.

**Comment by rudraya** [comment]
This issue was addressed in 6e99fab, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
### Update

This issue appears to have been fixed when testing in commit bb8e82dbcc. 

The following warning messages are given to show that the SDEID were removed due to duplicate TIME with different CONC.

```
Warning messages:
1: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '257914734' due to duplicate TIME but different CONC values
2: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '908214575' due to duplicate TIME but different CONC values
3: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '949633293' due to duplicate TIME but different CONC values
4: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'AMOUNTU' #2# is not present in the dataset provided via 'map'
5: In unit_conversion(data = data_data, map = map_data, result = computation_df,  :
  'PKPTMU' and 'CONCU' #10# are not present in the dataset provided via 'map'
```

The output for these SDEID is as follows:

```
> r[r$SDEID %in% c(257914734, 908214575, 949633293), ]
       SDEID DOSE DOSEC CMAX CMIN CLAST CEST CMAXDN TMAX TMIN TLAST TLAG KEL KELTMLO KELTMHI KELNOPT KELRSQ KELRSQA THALF THALFF LASTTIME
7  257914734   NA    NA   NA   NA    NA   NA     NA   NA   NA    NA   NA  NA      NA      NA      NA     NA      NA    NA   <NA>       NA
40 908214575   NA    NA   NA   NA    NA   NA     NA   NA   NA    NA   NA  NA      NA      NA      NA     NA      NA    NA   <NA>       NA
46 949633293   NA    NA   NA   NA    NA   NA     NA   NA   NA    NA   NA  NA      NA      NA      NA     NA      NA    NA   <NA>       NA
   AUCALL AUCALLDN AUCLAST AUCLASTC AUCLASTDN AUMCLAST AUC1 AUC2 AUC3 AUC4 AUC5 AUC6 AUC7 AUC8 AUC9 AUC10 AUC11 AUC12 AUC13 AUCINT1 AUCINT2
7      NA       NA      NA       NA        NA       NA   NA   NA   NA   NA   NA   NA   NA   NA   NA    NA    NA    NA    NA    <NA>    <NA>
40     NA       NA      NA       NA        NA       NA   NA   NA   NA   NA   NA   NA   NA   NA   NA    NA    NA    NA    NA    <NA>    <NA>
46     NA       NA      NA       NA        NA       NA   NA   NA   NA   NA   NA   NA   NA   NA   NA    NA    NA    NA    NA    <NA>    <NA>
   AUCINT3 AUCINT4 AUCINT5 AUCINT6 AUCINT7 AUCINT8 AUCINT9 AUCINT10 AUCINT11 AUCINT12 AUCINT13 AUC0_7 AUC8_48 AUC0_64 AUCINFO AUCINFP
7     <NA>    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>     <NA>     <NA>     <NA>     <NA>     NA      NA      NA      NA      NA
40    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>     <NA>     <NA>     <NA>     <NA>     NA      NA      NA      NA      NA
46    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>    <NA>     <NA>     <NA>     <NA>     <NA>     NA      NA      NA      NA      NA
   AUCINFOC AUCINFPC AUCINFODN AUCINFPDN AUMCINFO AUMCINFP MRTLAST MRTEVIFO MRTEVIFP AUCXPCTO AUCXPCTP AUMCXPTO AUMCXPTP CLFO CLFOW CLFP
7        NA       NA        NA        NA       NA       NA      NA       NA       NA       NA       NA       NA       NA   NA    NA   NA
40       NA       NA        NA        NA       NA       NA      NA       NA       NA       NA       NA       NA       NA   NA    NA   NA
46       NA       NA        NA        NA       NA       NA      NA       NA       NA       NA       NA       NA       NA   NA    NA   NA
   CLFPW VZFO VZFOW VZFP VZFPW CONC1 CONC2 CONC3 CONC4 CONC5 CONC6 CONC7 CONC8 CONC9 CONC10 CONC11 CONC12 CONC13 CONC14 CONCTIME1 CONCTIME2
7     NA   NA    NA   NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA     NA     NA     NA     NA     NA        NA        NA
40    NA   NA    NA   NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA     NA     NA     NA     NA     NA        NA        NA
46    NA   NA    NA   NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA     NA     NA     NA     NA     NA        NA        NA
   CONCTIME3 CONCTIME4 CONCTIME5 CONCTIME6 CONCTIME7 CONCTIME8 CONCTIME9 CONCTIME10 CONCTIME11 CONCTIME12 CONCTIME13 CONCTIME14 TIMEU
7         NA        NA        NA        NA        NA        NA        NA         NA         NA         NA         NA         NA    HR
40        NA        NA        NA        NA        NA        NA        NA         NA         NA         NA         NA         NA    HR
46        NA        NA        NA        NA        NA        NA        NA         NA         NA         NA         NA         NA    HR
   AMOUNTU DOSEU VOLUMEU CONCU KELU  CLU     AUCU       AUMCU    AUCNORMU AURCU CONCNORMU RATEU VOLUMEWU     CLWU    STUDY SITEID   SUBJID
7       NA    MG       L NG/ML 1/HR L/HR NG.HR/ML NG.HR.HR/ML NG.HR/ML/MG    NA  NG/ML/MG    NA     L/KG ML/HR/KG B1261002   1001 10011005
40      NA    MG       L NG/ML 1/HR L/HR NG.HR/ML NG.HR.HR/ML NG.HR/ML/MG    NA  NG/ML/MG    NA     L/KG ML/HR/KG B1261002   1001 10011002
46      NA    MG       L NG/ML 1/HR L/HR NG.HR/ML NG.HR.HR/ML NG.HR/ML/MG    NA  NG/ML/MG    NA     L/KG ML/HR/KG B1261002   1001 10011008
   RAND TREATXT TRTCD PKCOLL PKBDFLD      PKTERM PERIODU PERIOD VISITU VISIT                     ACTTRT ACTTRTC       DOSEP DOSPND ROUTE
7     5    3 mg     B  POINT  PLASMA PF-04634817  PERIOD      2    DAY     1  PF-04634817 3 MG SOLUTION       B PF-04634817   DONE  Oral
40    2   10 mg     C  POINT  PLASMA PF-04634817  PERIOD      3    DAY     1 PF-04634817 10 MG SOLUTION       C PF-04634817   DONE  Oral
46    8    3 mg     B  POINT  PLASMA PF-04634817  PERIOD      2    DAY     1  PF-04634817 3 MG SOLUTION       B PF-04634817   DONE  Oral
   RACES SEX  HT WT HTUNI WTUNI
7  WHITE   M 162 73    CM    KG
40 BLACK   M 176 93    CM    KG
46 BLACK   M 181 83    CM    KG
```
